src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -335,32 +349,38 @@ private void EventLoop()
                     Debug.Assert(numEvents > 0, $"Unexpected numEvents: {numEvents}");
 
                     bool enqueuedEvent = false;
-                    foreach (var socketEvent in new ReadOnlySpan<Interop.Sys.SocketEvent>(buffer, numEvents))
+                    lock (_handleToContextMap)
```

Comment:
This isn't a perf issue when there are multiple epoll threads?  That surprises me.

Reply:
Each epoll thread has it's own `_handleToContextMap`. It's accessed for a write only when a new socket is allocated or de-allocated (it's rare).  This loop is hot - it's being executed for every epoll notification.

From my initial observations that I am going to confirm with a lot of numbers is that:

* the perf is similar for x64 with not many connections (TechEmpower)
* the perf is better fo x64 with MANY connections
* the perf is better for ARM

Reply:
> Each epoll thread has it's own _handleToContextMap

Ah. Ok.

> From my initial observations that I am going to confirm with a lot of numbers is that

To my knowledge none of the TE benchmarks open/close sockets at any meaningful frequency.  But we know that happens in some real-world workloads, e.g. `Connection: close` requests.  It'd be good to validate that such use is not negatively impacted.

Reply:
> To my knowledge none of the TE benchmarks open/close sockets at any meaningful frequency. But we know that happens in some real-world workloads, e.g. Connection: close requests. It'd be good to validate that such use is not negatively impacted.

@antonfirsov @sebastienros do you have any idea how we could test this scenario?

Reply:
There's a benchmark for this in our infrastructure cc @sebastienros

Reply:
We do test that scenario actually by setting a "connection: close" header on the client side. We need to reduce the rate however from the defaults as we reach ports (or sockets) exhaustion quicly by doing that. We might need to find a better methodology if you have an idea for this. We can also build a simple console app that tests that and run it as a client if we know how to do it correctly instead of "connection: close".

Reply:
We even do this on https to exercise handshakes.

Reply:
To reduce the time the lock is held, the lookup could be performed up-front and `SocketAsyncContextWrapper` placed in an array of `EventBufferCount` size.